Added menhir_flags to dune_env.

bschommer opened the pull request:
As said in #2924 I would like to add possibility to add menhir flags to dune files. I started having a look at the source code but I'm currently unsure what places need to be touched. Any help would be appreciated.

bschommer committed:
Added menhir_flags to dune_env.

Signed-off-by: Bernhard Schommer <bernhardschommer@gmail.com>

nojb commented:
You need to
1) add `menhir_flags` to `Env_node.t`
2) add `Super_context.menhir_flags` to access them (a `Super_context.t` "carries" the necessary `Env_node.t`)
3) use `Super_context.menhir_flags` from `Menhir` when setting up the `menhir` rules.

I think you can look at `foreign_flags` for inspiration.

bschommer commented:
Thanks for the hinds, I will take a look.

bschommer committed:
Add menhir_flags to the rest.

Following foreign_flags menhir_flags are added to dune_env,
env_node and supercontext, as well as the expanding in menhir.

Signed-off-by: Bernhard Schommer <bernhardschommer@gmail.com>

rgrinberg commented:
Looking good. To speed up accepting your PR, could you:

* Run `$ make fmt` with the latest version of ocamlformat
* Add a test for this feature
* Update the documentation
* Update change log.

bschommer committed:
Formatted source code with make format.

Signed-off-by: Bernhard Schommer <bernhardschommer@gmail.com>

rgrinberg committed:
Simplify some code

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg committed:
Split menhir tests

This allows us to easily add more menhir tests

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg committed:
Add test for printing menhir flags in env

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg committed:
Print menhir flags with printenv

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg committed:
bump feature to 2.2

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

rgrinberg commented:
I added the last finishing touches.

@nojb @diml one thing that I'm unsure about is which language to use to version this. The env stanza belongs to the dune language, but this is technically functionality for the menhir plugin. What's the correct way to version this?

rgrinberg committed:
Document menhir_flags

Signed-off-by: Rudi Grinberg <me@rgrinberg.com>

bschommer commented:
Thanks for adding the tests and the last touches, I was away during the holiday season and did not have time to take a deeper look into this.

diml commented:
It should be versioned by the menhir language. In particular, you can't use `(menhir_flags ...)` inside `(env ...)` unless you have `(using menhir x.y)` in your `dune-project` file.

bschommer commented:
I rebased the branch, but I'm not sure how to change the dependency to the menhir version instead of the the dune language.

diml commented on src/dune/dune_env.ml:
```diff
@@ -14,6 +14,13 @@ module Stanza = struct
+  let menhir_flags ~since =
+    let check =
+      Option.map since ~f:(fun since ->
+          Dune_lang.Syntax.since Stanza.syntax since)
```
```suggestion
          Dune_lang.Syntax.since Menhir.syntax since)
```

So that the versioning is tied to `(using menhir 2.1)` rather than `(lang dune 2.2)`. You also need to add a 2.1 version to the menhir extension by replacing `(2, 0)` by `(2, 1)` in the definition of `Menhir.syntax`.

`Menhir.syntax` is defined inside `dune_file.ml` BTW

bschommer replied:
Thanks for the hints, I changed to `menhir`.

bschommer replied:
Unfortunately this introduces a dependency cycle, but I'm not sure how to resolve this.

diml replied:
Can it be resolved by moving the definition in the file? Here `Menhir` should refer to the `Menhir` sub-module in `dune_file.ml` rather than the toplevel `Menhir` module

bschommer replied:
I tried using `Dune_file.Menhir.syntax` instead and it did not work. I suspect that one needs to move `Menhir.syntax` into an own file like it is done for `Stanza.syntax`.

diml replied:
Ah yes, of course. I thought we were in `dune_file.ml` for some reason. Moving `Menhir.syntax` outside of `dune_file.ml` seems right. In fact, we can move the whole `Menhir` sub-module to its own file. I'm wondering if we can't just move that code to the toplevel `Menhir` module actually :thinking:

bschommer replied:
Moving the `Menhir` stuff into an own file does not really help, also moving `syntax` into `menhir.ml` does work.

bschommer replied:
In both cases I still got the cyclic dependency.

diml replied:
I had a look and I see why. Ideally we would make `Dune_env` extensible in a generic way so that its code doesn't mention menhir at all. For the time being, you can indeed just move only the `syntax` value to a dedicated file.

bschommer replied:
I moved the definition into a new file `menhir_stanza` and changed the usages to use the correct one.

diml commented:
@bschommer I left a comment in the code for that

bschommer committed:
Changed versioning.

The menhir flags feature should depend on the menhir language
instead of the dune language.

Signed-off-by: Bernhard Schommer <bernhardschommer@gmail.com>

bschommer committed:
Moved Menhir.syntax in own file.

Signed-off-by: Bernhard Schommer <bernhardschommer@gmail.com>

diml commented:
In the various test that fails, could you mask the menhir version (using a similar sed command as the one we use for `(lang dune <version>)`)? Apart from that it looks ready to me.

bschommer committed:
Fix version numbers in tests.

Signed-off-by: Bernhard Schommer <bernhardschommer@gmail.com>

diml commented:
This all looks good, thanks for your contribution! And thanks for splitting the menhir tests BTW :)

bschommer commented:
> This all looks good, thanks for your contribution! And thanks for splitting the menhir tests BTW :)

The tests were splitted by @rgrinberg .
